graylog2-server/src/main/java/org/graylog2/plugin/security/Permission.java

```diff
@@ -16,27 +16,135 @@
+     * <p>
+     * This can be use in Shiro permission checks, e.g. {@code isPermitted("users:edit", userId)'}.
+     *
+     * @return the permission string
+     */
+    String permission();
+
+    /**
+     * Returns a human-readable description of the permission.
+     *
+     * @return the description of the permission
+     */
+    String description();
+
+    /**
+     * Returns a map of GRN types to capabilities that this permission applies to.
+     * <p>
+     * This allows for specifying which GRN types this permission is applicable to and what capabilities are granted.
+     *
+     * @return an immutable map of GRN types to capabilities
+     */
+    ImmutableMap<GRNType, Capability> grnTypeCapabilities();
+
+    /**
+     * Converts this permission to a Shiro permission object using the specified GRN target.
+     *
+     * @param target the GRN target for which the permission is being checked
+     * @return a Shiro permission object representing this permission
+     */
+    org.apache.shiro.authz.Permission toShiroPermission(GRN target);
+
+    /**
+     * Creates a new permission with the specified permission string and description.
+     * <p>
+     * This is a convenience method to create a permission without GRN type capabilities.
+     *
+     * @param permission  the permission string, e.g. "users:edit"
+     * @param description a human-readable description of the permission
+     * @return a new Permission instance
+     */
+    static Permission create(String permission, String description) {
+        return ObjectActionPermission.create(permission, description, ImmutableMap.of());
+    }
+
+    /**
+     * Creates a new permission with the specified permission string, description, and GRN type capabilities.
+     * <p>
+     * This allows for specifying which GRN types and capabilities this permission applies to.
+     *
+     * @param permission          the permission string, e.g. "users:edit"
+     * @param description         a human-readable description of the permission
+     * @param grnTypeCapabilities a variable number of GRN type capabilities
+     * @return a new Permission instance
+     */
+    static Permission create(String permission, String description, GRNTypeCapability... grnTypeCapabilities) {
+        return ObjectActionPermission.create(
+                permission,
+                description,
+                Arrays.stream(grnTypeCapabilities)
+                        .collect(ImmutableMap.toImmutableMap(GRNTypeCapability::grnType, GRNTypeCapability::capability))
+        );
+    }
+
+    /**
+     * Creates a new {@link Capability#VIEW} capability mapping for the given GRN type.
+     * <p>
+     * Used for specifying which GRN types and capabilities a permission applies to.
+     *
+     * @param grnType the GRN type
+     * @return a GRNTypeCapability instance
+     */
+    static GRNTypeCapability addToViewCapabilityFor(GRNType grnType) {
```

Comment:
As discussed, the `addTo...` implies that this method does more than just creating a `GRNTypeCapability` object. One idea was to use something like a builder pattern and have methods like
```
Permission withViewCapabilityFor(GRNType grnType)
```
 which mutate the permission object.

Reply:
We discussed this offline and I changed the permission creation to a builder-like pattern using `.with<Capability>CapabilityFor(GRNType)` in 6700c20d7a.